### DISCUSSION TITLE
4 Wheeled Mobil Robot Using Lidar Navigation

### DISCUSSION THREAD

**Discussion by Qinssup** [fjp/2wd-robot#83] [General]
hi im so sorry, i am beginner in this field, really first time to build a robot 4wheel using lidar, but my final year project needed me to build this robot. im very lost, i have rp lidar a1, rasp pi 4, 4 dc motor, also have motor drive l298n, arduino since my first step is build robot obstacle avoidance using ultrasonic, then i need to do improvement from that. i did many research but i dont clearly understand, i installed Ubuntu 20.04, but i dont understand some videos tutorial. did u guys know what first step i need to do? i am lost

**Comment by fjp** [comment]
Hi @Qinssup, thanks for your question and nice project you are about to do. You will learn a lot :)

To understand the process of building a robot and what parts (hardware and software) are relevant, I suggest to follow the [ros-mobile-robots documentation](https://ros-mobile-robots.com/). You can find the menu on top (or on the left if you are using a mobile browser) and follow the chapters from left to right (or top to bottom):

![image](https://github.com/ros-mobile-robots/diffbot/assets/1286618/9a9d56eb-1230-44a7-a05a-7f2f3aaaeb09)

If you are done with a chapter you can navigate to the next one either from the menu or down at the bottom:

![image](https://github.com/ros-mobile-robots/diffbot/assets/1286618/a0fba379-e2c5-40b5-b5f7-f5dc2e660220)

The most important chapters are 

-  [Getting Started](https://ros-mobile-robots.com/technical_requirements/). Here you will most likely need to model your own robot in URDF. See the [diffbot_description ROS package](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_description) or [remo_description ROS package](https://github.com/ros-mobile-robots/remo_description) as examples. This was one of my first step.
- From the getting started page check that you have all technical requirements
- Organize your components. You should be able to use the ones you mention. A list of the ones I use for diffbot and REMO are found [here](https://ros-mobile-robots.com/components/)
- [Hardware setup](https://ros-mobile-robots.com/hardware_setup/overview/) explains assembly of the real robot and [electrical wiring](https://ros-mobile-robots.com/hardware_setup/electronics/) of the components. You would just replace the Teensy microcontroller I use with your Ardunio. And the motor driver (REMO uses Adafruit Feather wing and Diffbot uses Grove i2c motor driver) with your l298n.
- [The processing units chapter](https://ros-mobile-robots.com/processing_units/) explains how the parts interact with each other and which ROS nodes will be running when you fully use the REMO software. For diffbot the ROS nodes picture would look a bit different because the motor driver is also running on the raspberry pi and not the microcontroller.
- An important step in the processing units chapter is also the [hardware interfaces section](https://ros-mobile-robots.com/processing_units/hardware-interfaces/) which tells you about GPIO pins, i2c setup and ROS serial setup 
- ROS Network setup is also important to communicate between your development pc/laptop and the raspberry pi on the robot.
- Then follows the ROS software chapter where you should be able to navigate to the sections that are relevant for you at the moment

![image](https://github.com/ros-mobile-robots/diffbot/assets/1286618/35485a51-ca16-4539-a1f7-6a6576687c78)

The ROS software packages I spent the most time with are

- Robot description
- Base hardware interface contains 
    - the low level script code running on your microcontroller (Ardunio)
    - the high level control software running on the Raspberry Pi
- ROS Control software package where you will have to adapt the [mobile_base_controller](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_control/config/diffbot_control.yaml#L23) to use four wheels, according to the [example in the ros controls documentation](http://wiki.ros.org/diff_drive_controller?distro=noetic#Skid_steer)
- Simulation is not a must but nice to have when you are beginning with ROS and see if for example your robot description is setup correctly
- Other ROS software packages such as navigation or SLAM take relatively less time because most of it is ready to be used thanks to the ROS ecosystem. You just need to adapt some parameters which are explained in the documentation.

In the Base hardware interface you will have to do the most tweaking. For example you will have to decide where you run the motor driver code (Arduino or Raspberry Pi), and then update the code for the motor driver. But this should be the only hard part.

On the Home site you also find other [useful resources](https://ros-mobile-robots.com/#references) that helped me to get familiar with ROS:

![image](https://github.com/ros-mobile-robots/diffbot/assets/1286618/3e3b1ed0-a4c4-43bf-a666-e5a54ef788fd)

However, I think all steps you need should be distilled in the documentation as well. If not then feel free to ask here :)

> I also want to mention [linorobot](https://github.com/linorobot/linorobot) where you probably have to do less tweaking, because it directly supports your motor driver. The main diffference is that linorobot uses its own kinematics controller and runs most software parts on the microcontroller, whereas this project leverages mostly ROS packages such as [ROS Control](http://wiki.ros.org/ros_control), which makes it easy to setup simulation. The only custom software parts are the code for the motor driver (part of the [base hardware interface ROS package](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base)).
